cachi2/core/package_managers/yarn/resolver.py

```diff
@@ -344,13 +343,46 @@ def _generate_purl_for_package(package: _ResolvedPackage, project: Project) -> s
     :param project: the project object to resolve the configured registry url and file paths
         for file dependencies.
     """
-    # registry url can be accessed in project.yarnrc
-    # paths for file dependencies are relative to project.source_dir
+    qualifiers = dict()
+    subpath = None
+
+    # package with NpmLocator doesn't need any qualifiers
+
+    if isinstance(package.locator, HttpsLocator):
+        qualifiers["download_url"] = package.locator.url
+        if package.checksum is not None:
+            qualifiers["checksum"] = f"sha512:{package.checksum}"
+
+    elif isinstance(package.locator, WorkspaceLocator):
+        project_path = project.source_dir
+        workspace_path = package.locator.relpath
+
+        repo = get_repo_id(project_path)
+
+        qualifiers["vcs_url"] = repo.as_vcs_url_qualifier()
+        subpath = str(workspace_path)
+
+    elif isinstance(package.locator, (FileLocator, LinkLocator, PortalLocator)):
+        project_path = project.source_dir
+        workspace_path = package.locator.locator.relpath
+        package_path = package.locator.relpath
+
+        normalized = project_path.join_within_root(workspace_path, package_path)
+
+        repo = get_repo_id(project_path)
+        qualifiers["vcs_url"] = repo.as_vcs_url_qualifier()
+        subpath = str(normalized.subpath_from_root)
+
+    elif isinstance(package.locator, PatchLocator):
+        # ignore patch locators
+        # the actual dependency that is patched is reported separately
+        # the patch itself will be reported via SBOM pedigree patches
+        pass
+
```

Comment:
Would it be worth it to add a fallthrough `else` branch which would raise an exception about an unknown locator? I know it's highly unlikely a new locator would emerge and we'd have risen an exception in the resolver by now, so I'm just wondering about it, we may as well dismiss that case.

Reply:
something like `assert_never` ?

Reply:
Yeah, in fact I think we can just shamelessly copy-paste the whole `else` clause from https://github.com/containerbuildsystem/cachi2/blob/e228b16cfd395b0839348f014b745f1d56d25ef3/cachi2/core/package_managers/yarn/resolver.py#L273